File: frontend/src/services/api/vision.ts
/**
 * VIGILORA AI — Vision AI API Client Service
 */

const API_BASE = 'http://127.0.0.1:8000/api/v1/vision';

export interface DetectedObject {
  label: string;
  confidence: number;
  bbox?: number[];
  category?: string;
}

export interface ThreatItem {
  type: string;
  severity: 'NORMAL' | 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  confidence: number;
  description: string;
}

export interface VisionAnalysisResponse {
  analysis_id: string;
  camera_name: string;
  sector: string;
  threat_level: 'NORMAL' | 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  threat_score: number;
  confidence: number;
  summary: string;
  incident_title: string;
  incident_description: string;
  detected_objects: DetectedObject[];
  threats: ThreatItem[];
  location_context: string;
  visual_observations: string[];
  recommended_actions: string[];
  image_url?: string;
  image_metadata?: Record<string, any>;
  timestamp: string;
  is_demo_mode: boolean;
  model_provider: string;
}

export interface VisionScenario {
  id: string;
  title: string;
  category: string;
  sector: string;
  camera_name: string;
  threat_level: string;
  threat_score: number;
  description: string;
  image_url: string;
}

export interface VisionIncidentRecord {
  id: string;
  camera_name: string;
  sector: string;
  threat_level: string;
  threat_score: number;
  confidence: number;
  summary: string;
  incident_title: string;
  incident_description?: string;
  detected_objects_json?: DetectedObject[];
  threats_json?: ThreatItem[];
  visual_observations_json?: string[];
  recommended_actions_json?: string[];
  image_url?: string;
  image_metadata_json?: Record<string, any>;
  is_demo_mode: boolean;
  created_at?: string;
}

export async function fetchVisionScenarios(): Promise<VisionScenario[]> {
  const res = await fetch(`${API_BASE}/scenarios`);
  if (!res.ok) throw new Error('Failed to fetch demo scenarios');
  return res.json();
}

export async function analyzeVisionFrame(
  file?: File,
  scenarioId?: string,
  cameraName?: string,
  sector?: string
): Promise<VisionAnalysisResponse> {
  const formData = new FormData();
  if (file) {
    formData.append('file', file);
  }
  if (scenarioId) {
    formData.append('scenario_id', scenarioId);
  }
  if (cameraName) {
    formData.append('camera_name', cameraName);
  }
  if (sector) {
    formData.append('sector', sector);
  }

  const res = await fetch(`${API_BASE}/analyze`, {
    method: 'POST',
    body: formData,
  });

  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: 'Analysis failed' }));
    throw new Error(err.detail || 'Vision threat analysis failed');
  }

  return res.json();
}

export async function fetchVisionIncidents(): Promise<VisionIncidentRecord[]> {
  const res = await fetch(`${API_BASE}/incidents`);
  if (!res.ok) throw new Error('Failed to fetch vision incidents history');
  return res.json();
}

export async function fetchVisionIncident(id: string): Promise<VisionIncidentRecord> {
  const res = await fetch(`${API_BASE}/incidents/${id}`);
  if (!res.ok) throw new Error(`Failed to fetch vision incident ${id}`);
  return res.json();
}

export async function saveVisionIncident(
  data: Partial<VisionAnalysisResponse>
): Promise<VisionIncidentRecord> {
  const res = await fetch(`${API_BASE}/incidents`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });

  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: 'Failed to save incident' }));
    throw new Error(err.detail || 'Failed to save incident');
  }

  return res.json();
}

File: backend/schemas/vision.py
"""Pydantic schemas for Vision AI Multimodal Threat Analysis."""

from datetime import datetime
from typing import List, Optional, Any
from pydantic import BaseModel, ConfigDict, Field


class DetectedObject(BaseModel):
    label: str
    confidence: float = Field(ge=0.0, le=1.0)
    bbox: Optional[List[int]] = None  # [x1, y1, x2, y2]
    category: Optional[str] = "object"  # person | vehicle | package | hazard | access


class ThreatItem(BaseModel):
    type: str  # e.g., "perimeter_breach", "suspicious_package", "crowd_anomaly", "fire_hazard"
    severity: str  # NORMAL | LOW | MEDIUM | HIGH | CRITICAL
    confidence: float = Field(ge=0.0, le=1.0)
    description: str


class VisionAnalysisResponse(BaseModel):
    analysis_id: str
    camera_name: str = "Camera 04 (Main Security)"
    sector: str = "Sector 7 (Perimeter)"
    threat_level: str  # NORMAL | LOW | MEDIUM | HIGH | CRITICAL
    threat_score: float  # 0.0 to 100.0
    confidence: float  # 0.0 to 1.0
    summary: str
    incident_title: str
    incident_description: str
    detected_objects: List[DetectedObject] = []
    threats: List[ThreatItem] = []
    location_context: str
    visual_observations: List[str] = []
    recommended_actions: List[str] = []
    image_url: Optional[str] = None
    image_metadata: Optional[dict] = None
    timestamp: str
    is_demo_mode: bool = False
    model_provider: str = "VIGILORA Vision Core"

    model_config = ConfigDict(protected_namespaces=())



class VisionIncidentCreate(BaseModel):
    analysis_id: str
    camera_name: str = "Camera 04 (Main Security)"
    sector: str = "Sector 7 (Perimeter)"
    threat_level: str = "HIGH"
    threat_score: float = 85.0
    confidence: float = 0.90
    summary: str
    incident_title: str = "Suspicious Visual Activity"
    incident_description: Optional[str] = None
    detected_objects: List[DetectedObject] = []
    threats: List[ThreatItem] = []
    visual_observations: List[str] = []
    recommended_actions: List[str] = []
    image_url: Optional[str] = None
    image_metadata: Optional[dict] = None
    is_demo_mode: bool = False


class VisionIncidentResponse(BaseModel):
    id: str
    camera_name: str
    sector: str
    threat_level: str
    threat_score: float
    confidence: float
    summary: str
    incident_title: str
    incident_description: Optional[str] = None
    detected_objects_json: Optional[Any] = None
    threats_json: Optional[Any] = None
    visual_observations_json: Optional[Any] = None
    recommended_actions_json: Optional[Any] = None
    image_url: Optional[str] = None
    image_metadata_json: Optional[Any] = None
    is_demo_mode: bool = False
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class VisionScenario(BaseModel):
    id: str
    title: str
    category: str
    sector: str
    camera_name: str
    threat_level: str
    threat_score: float
    description: str
    image_url: str